### PR TITLE
Created louvain communities action for community detection in networks

### PR DESCRIPTION
**Pull request by rhernandvel**
New _louvain.py script for method that allows to obtain consensus partition.
New sematic type: `Node_Community`
Updated README with python libraries needed for the Louvain method.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,6 +40,12 @@ In the same conda environment pip install from the q2-makarsa github repo:
 pip install git+https://github.com/BenKaehler/q2-makarsa.git
 ```
 
+Install the python libraries needed for the community prediction:
+
+```
+pip install python-louvain
+pip install netneurotools
```
Comment (nbokulich) [inline README.md:47]: neurotools is not conda installable (which will complicate installation). Can this be replaced by another package?

File: q2_makarsa/_network.py [hunk 2]
```diff
@@ -21,3 +21,18 @@ def _validate_(self, level):
 )
 
 Network = qiime2.plugin.SemanticType("Network")
+
+class NodeCommunityFormat(model.TextFileFormat):
+    def _validate_(self, level):
+        try:
+            table = pd.read_table(str(self))
+            if not isinstance(table, pd.DataFrame):
+                raise ValidationError("")
```
Comment (nbokulich) [inline q2_makarsa/_network.py:30]: please insert an error message here. Given that there is a separate `ParseError` I am guessing that this error would be caused by an empty table?

File: q2_makarsa/plugin_setup.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 from networkx import Graph, read_graphml, write_graphml
 from q2_types.feature_table import FeatureTable, Frequency
 from qiime2.plugin import Bool, Float, Int, Metadata, Plugin, Str
+import pandas as pd
 
-from ._network import Network, NetworkDirectoryFormat, NetworkFormat
+from ._network import Network, NetworkDirectoryFormat, NetworkFormat, NodeCommunityFormat, Node_Community, NodeDirectoryFormat
```
Comment (nbokulich) [inline q2_makarsa/plugin_setup.py:6]: NodeCommunity is rather long and not too intuitive to me. What about `NodeMap` or `NetworkModules` or something like this?

File: q2_makarsa/plugin_setup.py [hunk 4]
```diff
@@ -107,3 +108,42 @@ def _2(ff: NetworkFormat) -> Graph:
...
+plugin.register_semantic_types(Node_Community)
+plugin.register_formats(NodeCommunityFormat, NodeDirectoryFormat)
+plugin.register_semantic_type_to_format(
+    Node_Community, artifact_format=NodeDirectoryFormat
+)
+
+
+
+
+@plugin.register_transformer
```
Comment (nbokulich) [inline q2_makarsa/plugin_setup.py:121]: you also need a transformer to go from `NodeCommunityFormat` to `pd.DataFrame` and from `NodeCommunityFormat` to `qiime2.Metadata`. The latter will allow the `NodeCommunityFormat` artifact to be passed to other existing actions in QIIME2, e.g., for collapsing a table into modules.

**Review by nbokulich — CHANGES_REQUESTED** [review]
thanks @rhernandvel ! I have a few comments in-line. My main comment is that unit tests are missing and needed. Transformers are also needed to make this usable with downstream actions.

I have done a somewhat high-level review more on the plugin registration and semantics... @BenKaehler could you please also test and review the functions?

File: README.md (hunk 1 above)
Comment (nbokulich) [reply README.md:47]: @BenKaehler what is your input on this?

**Comment by BenKaehler** [comment]
Thank you very much @rhernandvel for this commit! It is very much appreciated.

Thanks also @nbokulich for your valuable comments.

I have a few high-level comments (that are probably a bit painful, sorry).

- We have worked hard to keep the dependencies down, so I'm reluctant to add python-louvain or netneurotools.
  - Louvain community detection is available in networkx (an existing dependency). Could we use that instead?
  - Clustering consensus is a bit harder, but could we use a [VotingClassifier](https://scikit-learn.org/stable/modules/ensemble.html#voting-classifier) from scikit-learn (also an existing dependency) to similar effect?
- I agree with @nbokulich that we need to be able to use the output of this algorithm. Can we reuse an existing semantic type?
  - Probably the easiest way it to just output another network with the labels annotated on the nodes. See the spiec-easi and flashweave functions for examples of how to do that.
  - You could reuse the Hierarchy core qiime2 type. [Here](https://github.com/qiime2/q2-gneiss/blob/c24f826687627dbe3115593ebf0550b650144a27/q2_gneiss/cluster/_cluster.py#L48) is an example.

The other thing is that it will probably take me a while to take full advantage of community information. Ultimately it would be great to be able to collapse large networks interactively, but for the first release I will probably target colouring nodes. An intermediate step might be a [hierarchical edge bundling](https://vega.github.io/vega/examples/edge-bundling/) visualisation. As such, an annotated network or any output that could be transformed to metadata would work in the first instance, and we could get fancier later.

Thanks again!

**Comment by BenKaehler** [comment]
After further thought, writing our own consensus clustering function would be a simple alternative to neurotools.

**Comment by rhernandvel** [comment]
Thanks for the comments @BenKaehler!

- Python-louvain will be replaced by the louvain community detection in networkx. Python-louvain was initially selected because it allows to obtain a dendogram which contains all the levels in a simpler output format compared to `nx.community` `louvain_partitions`. 

- Obtaining a consensus is desirable since it has been shown to lead to more accurate partitions compared to single runs of community detection methods [(Lancichinetti & Fortunato, 2012)](https://www.nature.com/articles/srep00336).

- For the output of the algorithm, I consider that using a table format instead the of a network with the labels annotated on the nodes would be more scalable and allow for several features to be visualized in the graph. The semantic type for this would still be `NodeMap` (currently named `NodeCommunityFormat`) and the transformers to go from `NodeMap` to `pd.DataFrame` and from `NodeMap` to `qiime2.Metadata` would be included as suggested by @nbokulich .

- Taking into account that the partitions with the highest modularity per run are the ones used to generate the consensus and that a method for visualization is not available for hierarchy it would not be necessary to obtain use the Hierarchy core qiime2 type.

- Lastly, I will also check if there is a clustering algorithm in the already present libraries that could be useful and otherwise write the consensus clustering function for this.

If all this sounds good, I will proceed to integrate these steps!

**Comment by BenKaehler** [comment]
Thanks @rhernandvel, that sounds great.

Will you include the full hierarchy in the node map table? You could separate levels using `"; "`, as we do for taxonomies. In fact, I would feel tempted to recycle `FeatureData[Taxonomy]` for the purpose. If you do, the visualiser will recognise the levels and allow them to be displayed without further modification. Maybe consult with @nbokulich on that one.

Regarding the consensus clustering algorithm, it looks like it would be pretty straightforward to implement. This is the one that they use in [Lancichinetti & Fortunato, 2012](https://www.nature.com/articles/srep00336):

> The starting point is a network  with n vertices and a clustering algorithm A.
>   1 Apply A on  nP times, so to yield nP partitions.
>   2 Compute the consensus matrix D, where Dij is the number of partitions in which vertices i and j of  are assigned to the same cluster, divided by nP.
>   3 All entries of D below a chosen threshold τ are set to zero.
>   4 Apply A on D nP times, so to yield nP partitions.
>   5 If the partitions are all equal, stop (the consensus matrix would be block-diagonal). Otherwise go back to 2.

I would be happy to contribute it if it comes to that.

Thanks again, I look forward to seeing your revisions.

**Comment by rhernandvel** [comment]
Thanks again for the comments @BenKaehler

I was thinking of only including the level with the highest modularity in the `NodeMap` table because the consensus is obtained only for this level. If the hierarchy is to be stored, separating the levels using  `“;”` seems a suitable option.   

The consensus clustering algorithm used in [Lancichinetti & Fortunato (2012)](https://www.nature.com/articles/srep00336)  is a good option and also straightforward to implement. The downside to this method is that it generates a consensus that may correspond in some cases to a new partition different from any of the nP partions, thus losing the higher levels (hierarchy).

Looking forward to seeing your comments on this

**Comment by BenKaehler** [comment]
Thanks @rhernandvel, I'm happy with you returning the level with the highest modularity. 

Does that solve the problems with the Lancichinetti & Fortunato algorithm?

Either way I'm happy with any consensus clustering algorithm you can implement or vendor.

**Comment by rhernandvel** [comment]
Great! Thanks @BenKaehler 
Yes, this solves the issue with the Lancichinetti & Fortunato algorithm.
I will then include all these changes.

**Comment by BenKaehler** [comment]
Hi @rhernandvel, thanks very much for your latest commit, I can't wait to try it.

Could you please tidy up a bit to remove the changes to the README and resolve the conflicts in plugin_setup.py?

I should be able to merge without conflicts and it should pass existing CI.

If you have time, some unit tests for _louvain.py would be fantastic, but we can sort that out later if you want.

Thanks!

**Comment by rhernandvel** [comment]
Hi @BenKaehler ,
I removed the changes to the README and resolve the conflicts in plugin_setup.py.
I also added a test for _louvain.py.
Thanks!

**Review by BenKaehler — APPROVED** [review]
Thanks @rhernandvel! Very much appreciated.

**Pull request merged**